exposure-notification/src/main/java/fi/thl/covid19/exposurenotification/efgs/FederationGatewayService.java

```diff
@@ -0,0 +1,125 @@
+package fi.thl.covid19.exposurenotification.efgs;
+
+import fi.thl.covid19.exposurenotification.diagnosiskey.DiagnosisKeyDao;
+import fi.thl.covid19.exposurenotification.diagnosiskey.IntervalNumber;
+import fi.thl.covid19.exposurenotification.diagnosiskey.TemporaryExposureKey;
+import fi.thl.covid19.exposurenotification.error.EfgsOperationException;
+import fi.thl.covid19.proto.EfgsProto;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Service;
+
+import java.time.Instant;
+import java.util.*;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static fi.thl.covid19.exposurenotification.efgs.FederationGatewayBatchUtil.*;
+
+@Service
+public class FederationGatewayService {
+
+    private static final int BATCH_MAX_SIZE = 5000;
+
+    private final FederationGatewayClient client;
+    private final DiagnosisKeyDao dd;
+    private final FederationOperationDao fod;
+    private final FederationBatchSigner signer;
+
+    public FederationGatewayService(
+            FederationGatewayClient client,
+            DiagnosisKeyDao diagnosisKeyDao,
+            FederationOperationDao fod,
+            FederationBatchSigner signer
+    ) {
+        this.client = client;
+        this.dd = diagnosisKeyDao;
+        this.fod = fod;
+        this.signer = signer;
+    }
+
+    public long startOutbound() {
+        long operationId = fod.startOutboundOperation();
+
+        if (operationId > 0) {
+            doOutbound(operationId);
+        }
+        return operationId;
+    }
+
+    public void startInbound(Optional<String> batchTag) {
+        String date = getDateString(Instant.now());
```

Comment:
If inbound ops trigger on both nodes at the same time, both fetch the same data, right? It's probably not a big issue, but we could use the status-table to only have one of these running at a time.

Reply:
This is a good point, however efgs maintains state for these, so this probably isn't a problem. But I haven't checked this yet in this level. However, with callback we are also on one node. This is a problem only with scheduled run.